### ISSUE TITLE
Duplicate primers when reading SnapGene file

### ISSUE THREAD

**Issue by nicolazilio** [biopython/biopython#4217]
### Setup

I am reporting a problem with Biopython version, Python version, and operating system as follows:

```Python
3.8.13 (default, Jul 24 2022, 18:39:26) 
[GCC 8.3.0]
CPython
Linux-4.19.0-23-amd64-x86_64-with-glibc2.28
1.80
```

### Expected behaviour
Equivalent plasmid maps in .dna and .gbk formats should contain the same features when parsed with the snapgene and genbank parsers.

### Actual behaviour

@gouttegd 

Using SnapGene Server, I save a plasmid map as .dna and .gbk. When I parse those files, which should be identical to each other(?), in Biopython, the SnapGene parser reports duplicated primers, while the GenBank file does not, as one would expect. 

For example

```Python
from Bio import SeqIO
from collection_management.models import Plasmid

for p in Plasmid.objects.filter(created_by__id=2, id=4403):

    print(f"*** {p.id} ***\n")

    sg_map = SeqIO.read(p.map.path, "snapgene")
    print('Snapgene')
    print('\t' + "\t".join(sorted(
        [f"{f.qualifiers['label'][0]} {f.location}\n"
         for f in sg_map.features
         if f.qualifiers.get('label', False)])))

    print('Genbank')
    gb_map = SeqIO.read(p.map_gbk.path, "genbank")
    print('\t' + "\t".join(sorted(
        [f"{f.qualifiers['label'][0]} {f.location}\n"
         for f in gb_map.features
         if f.qualifiers.get('label', False)])))
```

returns 

```
*** 4403 ***

Snapgene
	1468 [767:788](+)
	1468 [767:788](+)
	5076 [4340:4365](-)
	5076 [4340:4365](-)
	5093 [897:957](+)
	5093 [897:957](+)
	5094 [3192:3234](-)
	5094 [3192:3234](-)
	5095 [3194:3238](+)
	5095 [3194:3238](+)
	AmpR join{[7956:8025](-), [7164:7956](-)}
	AmpR promoter [8025:8130](-)
	CAP binding site [6096:6118](+)
	CMV enhancer [234:614](+)
	CMV promoter [614:818](+)
	CTCF [992:3173](+)
	CTCF [992:3173](+)
	Dam methylase [3251:4091](+)
	FLAG [3221:3248](+)
	FLAG [3224:3248](+)
	FRT [4626:4674](+)
	HygR [4681:5704](+)
	M13 rev [6003:6020](-)
	SV40 poly(A) signal [5833:5955](+)
	attB1 [962:987](+)
	attB2 [3177:3202](-)
	bGH poly(A) signal [4118:4343](+)
	lac operator [6027:6044](+)
	lac promoter join{[6076:6082](-), [6058:6076](-), [6051:6058](-)}
	ori [6405:6994](-)
	tet operator [819:838](+)
	tet operator [840:859](+)

Genbank
	1468 [768:789](+)
	5076 [4341:4366](-)
	5093 [898:958](+)
	5094 [3193:3235](-)
	5095 [3195:3239](+)
	AmpR [7164:8025](-)
	AmpR promoter [8025:8130](-)
	CAP binding site [6096:6118](+)
	CMV enhancer [234:614](+)
	CMV promoter [614:818](+)
	CTCF [992:3173](+)
	CTCF [992:3173](+)
	Dam methylase [3251:4091](+)
	FLAG [3221:3248](+)
	FLAG [3224:3248](+)
	FRT [4626:4674](+)
	HygR [4681:5704](+)
	M13 rev [6003:6020](-)
	SV40 poly(A) signal [5833:5955](+)
	attB1 [962:987](+)
	attB2 [3177:3202](-)
	bGH poly(A) signal [4118:4343](+)
	lac operator [6027:6044](+)
	lac promoter [6051:6082](-)
	ori [6405:6994](-)
	tet operator [819:838](+)
	tet operator [840:859](+)

```

The plasmid maps are attached.

Thanks!
[pHU4403 - pDEST_FRT_TO-Cterm-FLAG-Dam - CTCF.zip](https://github.com/biopython/biopython/files/10455588/pHU4403.-.pDEST_FRT_TO-Cterm-FLAG-Dam.-.CTCF.zip)

**Comment by peterjc** [comment]
The lac promoter appears to be fragmented into three contiguous pieces in the snapgene file, that's also odd.

@gouttegd Are you able to take a look at this please?

**Comment by gouttegd** [comment]
@peterjc Yes, you may assign to me.

**Comment by nicolazilio** [comment]
> The lac promoter appears to be fragmented into three contiguous pieces in the snapgene file, that's also odd.
> 
> @gouttegd Are you able to take a look at this please?

The slit feature is actually correct, see https://github.com/biopython/biopython/issues/3655#issuecomment-884945935

**Comment by gouttegd** [comment]
It seems the SnapGene folks have come up with a new idea…

All primer specifications in that file have _two_ `BindingSite` nodes instead of one:

```xml
<Primer recentID="4" name="1468" sequence="CGCAAATGGGCGGTAGGCGTG">
  <BindingSite location="768-788" boundStrand="0" annealedBases="CGCAAATGGGCGGTAGGCGTG" meltingTemperature="65">
    <Component hybridizedRange="768-788" bases="CGCAAATGGGCGGTAGGCGTG"/>
  </BindingSite>
  <BindingSite simplified="1" location="768-788" boundStrand="0" annealedBases="CGCAAATGGGCGGTAGGCGTG" meltingTemperature="65">
    <Component hybridizedRange="768-788" bases="CGCAAATGGGCGGTAGGCGTG"/>
  </BindingSite>
</Primer>
```

For each primers, the two `BindingSite` nodes are completely identical except for the `simplified="1"` attribute on the second node. Not sure what they were thinking… Looks like the second node was intended to be a “simplified” representation of the primer’s binding site, but then they forgot to actually simplify it… :D 

I’ll update the parser to simply ignore these so-called “simplified” primers if a “non-simplified” one already exists.

**Comment by peterjc** [comment]
Do either of you have a technical contact at Snapgene to raise this issue with? It would be even nicer to just have them fix their output.

**Comment by nicolazilio** [comment]
I can contact them. But I suspect they'll just say that's how it's supposed to be.

**Comment by peterjc** [comment]
Yes please @nicolazilio - and if they say it is deliberate, the @gouttegd's proposal sounds good.

**Comment by gouttegd** [comment]
I just noticed that the example output given in this issue highlights another problem:

```
Snapgene
	1468 [767:788](+)
	5076 [4340:4365](-)
[...]
Genbank
	1468 [768:789](+)
	5076 [4341:4366](-)
```

All primers’ coordinates obtained from the SnapGene parser are off by -1. Looking at the SnapGene file, those are the coordinates it contains, it’s not an error from the parser.

I did some test with SnapGene Viewer:

* If I create any kind of feature with a range of, say, 20-50, the feature is written down in the SnapGene file with the coordinates `20-50`, as expected.
* If I create a primer with the same 20-50 range, the primer is written down in the `Primers` packet of the SnapGene file with the coordinates `19-49`.

So SnapGene shifts all primers coordinates by -1 when it writes them down, and re-shifts them by +1 when reading them in order to correctly display the intended range.

I have no idea why it does something like that, but our parser must take it into account by parsing primers’ coordinates differently from features’ coordinates (applying a +1 correction only to the formers).

**Comment by nicolazilio** [comment]
Well spotted! 

I wonder whether it would be a good idea to implement a test that compares the outputs (name, start and end of features?) of Biopython's SnapGene and GenBank parsers with .dna and .gb maps generated from the same plasmid in SnapGene?

**Comment by gouttegd** [comment]
On second thought, I don’t think such a test would be a good idea.

SnapGene can’t even be bothered to write fully correct GenBank files, it generates an incorrect LOCUS line (not incorrect enough to prevent Biopython’s GenBank parser to read the file, but still enough that the parser complains about it). I don’t like the idea of using files of dubious quality as reference in the test suite.

What we currently do is that the expected names and locations of each feature in our handful of test files are hardcoded in the SnapGene test case. The expected locations were taken “manually“ by opening the test files in SnapGene and checking what the coordinates should be. This should be enough.

This was _not_ enough to catch the issue with primers, but this is only because until now none of our test files had any primers (an unfortunate oversight of my part when I created those test files). Even if had had a test fixture such as the one envisioned above, without a sample file containing primers, it would _not_ have detected the problem.

**Comment by peterjc** [comment]
I'm happy to leave this to your judgement as our SnapGene expert. Elsewhere we do have a mixture of dual format comparison and standalone test cases. Both approaches have their merits.

**Comment by peterjc** [comment]
This fix will be in Biopython 1.81 which I'm hoping will be released next month (Feb 2023). Thank you both.
